### PR TITLE
Process updates to ServiceAttachment spec

### PR DESCRIPTION
**Pull request by swetharepakula**
* all validation will be left to the GCE Service Attachment API

### CODE REVIEW

**Comment by swetharepakula** [comment]
Rebased this PR on the current master.

The `needsUpdate` function requires that all updates to the spec are checked field by field, therefore to catch when a new field is added, I have added a unit test. The purpose is to inform developers that they need to update `needsUpdate` when a new field is added to the spec.

**Comment by swetharepakula** [comment]
/hold

PR needs to be updated to include the new fields (ConsumerAllowList and ConsumerRejectList)

File: pkg/psc/controller.go [hunk 1]
```diff
@@ -575,52 +583,57 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
 }
 
-// validateUpdate will validate whether ServiceAttachment matches the GCE Service Attachment
-// resource. If not, validateUpdate will return an error, since GCE Service Attachments cannot
-// be updated after creation
-func validateUpdate(existingSA, desiredSA *beta.ServiceAttachment) error {
+// needsUpdate will determine whether ServiceAttachment matches the GCE Service Attachment
+// resource. If not, needsUpdate will return true. needsUpdate will not validate whether
+// the update will be successful or not
+func needsUpdate(existingSA, desiredSA *beta.ServiceAttachment) (bool, error) {
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go:599]: I took some of this approach. After checking the forwarding rules and subnets, we can set the values on the copy of the desiredSA and do a deep equal with the rest of the fields.

The only other change needed to be able to do the deep equal is generating the desiredSA object based on the existingSA.

**Comment by swetharepakula** [comment]
/hold cancel

The PR has been rebased and the update check has been updated as well. This PR can be reviewed now.

File: pkg/psc/controller_test.go [hunk 2]
```diff
@@ -436,41 +436,34 @@ func TestServiceAttachmentConsumers(t *testing.T) {
...
 	svcName := "my-service"
+	otherServiceName := "my-other-service"
 	saUID := "serivce-attachment-uid"
 	frIPAddr := "1.2.3.4"
 
-	saCRAnnotation := testServiceAttachmentCR(saName, svcName, saUID, []string{"my-subnet"}, true)
+	saCRAnnotation := testServiceAttachmentCR(saName, svcName, saUID, []string{"my-subnet"}, true, false)
 	saCRAnnotation.Annotations = map[string]string{"some-key": "some-value"}
 
 	testcases := []struct {
```
Comment (prameshj) [inline pkg/psc/controller_test.go:446]: Can you add a testcase where the existing and desired SA have the same number of subnets but differ in just 1 value?

File: pkg/psc/controller.go [hunk 3]
```diff
@@ -314,41 +314,56 @@ func (c *Controller) processServiceAttachment(key string) error {
...
 		if err != nil {
-			return fmt.Errorf("invalid Service Attachment Update: %w", err)
+			return fmt.Errorf("unable to process Service Attachment Update: %w", err)
+		}
+
+		if shouldUpdate {
+			// In order for the update to be successful, the self link in the target service must be
+			// exactly the same. needsUpdate throws an error in situations the forwarding rule was
+			// changed on the spec, so only updates where the rule is the same will be updated
+			gceSvcAttachment.TargetService = existingSA.TargetService
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go:361]: The issue is that the selflink changes based on the resource version used (GA, Beta, Alpha) and it does not depend on what we set but what the GCE API does. For example when it was in alpha the selflinks were always in alpha and now they are in beta. And there is a validation error if the selflink is not exactly the same, which is the reason we have to set it. We cannot set it earlier in order to catch when the user tries to change the target service (which will fail).

I have added a similar explanation in the coment.

File: pkg/psc/controller.go [hunk 4]
```diff
@@ -500,7 +515,7 @@ func (c *Controller) getSubnetURLs(subnets []string) ([]string, error) {
 // updateServiceAttachmentStatus updates the CR's status with the GCE Service Attachment URL
 // and the producer forwarding rule
 func (c *Controller) updateServiceAttachmentStatus(cr *sav1beta1.ServiceAttachment, gceSAKey *meta.Key) (*sav1beta1.ServiceAttachment, error) {
-	gceSA, err := c.cloud.Compute().AlphaServiceAttachments().Get(context2.Background(), gceSAKey)
+	gceSA, err := c.cloud.Compute().BetaServiceAttachments().Get(context2.Background(), gceSAKey)
 	if err != nil {
 		return cr, fmt.Errorf("failed to query GCE Service Attachment: %w", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: done

File: pkg/psc/controller.go [hunk 5]
```diff
@@ -600,52 +615,68 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
-// validateUpdate will validate whether ServiceAttachment matches the GCE Service Attachment
-// resource. If not, validateUpdate will return an error, since GCE Service Attachments cannot
-// be updated after creation
-func validateUpdate(existingSA, desiredSA *beta.ServiceAttachment) error {
-	if existingSA.ConnectionPreference != desiredSA.ConnectionPreference {
-		return fmt.Errorf("serviceAttachment connection preference cannot be updated from %s to %s", existingSA.ConnectionPreference, desiredSA.ConnectionPreference)
-	}
+// needsUpdate will determine whether ServiceAttachment matches the GCE Service Attachment
+// resource. If not, needsUpdate will return true. needsUpdate will not validate whether
+// the update will be successful or not
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed.

File: pkg/psc/controller.go [hunk 6]
```diff
@@ -600,52 +615,68 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
+func needsUpdate(existingSA, desiredSA *beta.ServiceAttachment) (bool, error) {
+	// NOTE: The selflinks cannot be directly compared as the selflink we generate may not
+	// be the same as the one that eventually gets stored on the GCE object. For example
+	// the controller takes the forwarding rule from the GA FR resource, however if the GCE
+	// SA uses the Beta FR resource the self links will be different though the resource is
+	// the same. The same is true for the subents. Due to this discrepency the GCE SA cannot
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed

File: pkg/psc/controller.go [hunk 7]
```diff
@@ -600,52 +615,68 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
+	// be compared with a reflect.DeepEqual.
 
 	// The TargetService on the GCE Service Attachment is the self link to the URL of the producer
 	// forwarding rule (L4 ILB).
 	existingFR, err := cloud.ParseResourceURL(existingSA.TargetService)
 	if err != nil {
-		return fmt.Errorf("serviceAttachment existing target service has malformed URL: %w", err)
+		return false, fmt.Errorf("serviceAttachment existing target service has malformed URL: %w", err)
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed here and below.

File: pkg/psc/controller.go [hunk 8]
```diff
@@ -600,52 +615,68 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
 	}
 	if !reflect.DeepEqual(existingFR, desiredFR) {
-		return fmt.Errorf("serviceAttachment target service cannot be updated from %s to %s", existingSA.TargetService, desiredSA.TargetService)
+		return true, fmt.Errorf("serviceAttachment forwarding rule cannot be updated ")
 	}
 
 	if len(existingSA.NatSubnets) != len(desiredSA.NatSubnets) {
-		return fmt.Errorf("serviceAttachment NAT Subnets cannot be updated")
+		return true, nil
 	} else {
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: thanks for the suggestion! made the function much easier to read :)

File: pkg/psc/controller.go [hunk 9]
```diff
@@ -314,41 +314,56 @@ func (c *Controller) processServiceAttachment(key string) error {
...
+		// Most of the validation is left to the GCE Service Attachment API. needsUpdate only checks
+		// to see if the spec has changed and whether an update is necessary.
+		shouldUpdate, err := needsUpdate(existingSA, gceSvcAttachment)
 		if err != nil {
-			return fmt.Errorf("invalid Service Attachment Update: %w", err)
+			return fmt.Errorf("unable to process Service Attachment Update: %w", err)
+		}
+
+		if shouldUpdate {
+			// In order for the update to be successful, the self link in the target service must be
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: Good call, I clarified in the comment that target service is the same as the forwarding rule

File: pkg/psc/controller_test.go (hunk 2 above)
Comment (swetharepakula) [reply pkg/psc/controller_test.go:450]: Added a test case. The test case will pass in the tests, but in reality GCE API will fail the update. I added a comment about the same in the tests.

File: pkg/psc/controller.go [hunk 10]
```diff
@@ -314,41 +314,59 @@ func (c *Controller) processServiceAttachment(key string) error {
...
+	desc := sautils.NewServiceAttachmentDesc(updatedCR.Namespace, updatedCR.Name, c.clusterName, c.clusterLoc, c.regionalCluster)
+	consumerAcceptLists := convertAllowList(updatedCR.Spec)
+	gceSvcAttachment.ConnectionPreference = svcAttachment.Spec.ConnectionPreference
+	gceSvcAttachment.Name = saName
+	gceSvcAttachment.NatSubnets = subnetURLs
+	gceSvcAttachment.TargetService = frURL
+	gceSvcAttachment.Region = c.cloud.Region()
+	gceSvcAttachment.Description = desc.String()
+	gceSvcAttachment.EnableProxyProtocol = updatedCR.Spec.ProxyProtocol
+	gceSvcAttachment.ConsumerAcceptLists = consumerAcceptLists
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed

File: pkg/psc/controller.go [hunk 11]
```diff
@@ -314,41 +314,59 @@ func (c *Controller) processServiceAttachment(key string) error {
...
 		if err != nil {
-			return fmt.Errorf("invalid Service Attachment Update: %w", err)
+			return fmt.Errorf("unable to process Service Attachment Update: %w", err)
+		}
+
+		if shouldUpdate {
+			// In order for the update to be successful, the self link in the target service (same resource
+			// as the forwarding rule) must be exactly the same. needsUpdate throws an error in situations
+			// the forwarding rule/targetservice was changed on the spec, so only updates where the rule is
+			// the same will be updated PSC controller generates the selflink of GA version of the resource,
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed.

File: pkg/psc/controller.go [hunk 12]
```diff
@@ -600,52 +618,67 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
+		return false, fmt.Errorf("serviceAttachment existing target service URL, %q, is malformed: %w", existingSA.TargetService, err)
 	}
 	desiredFR, err := cloud.ParseResourceURL(desiredSA.TargetService)
 	if err != nil {
-		return fmt.Errorf("serviceAttachment desired target service has malformed URL: %w", err)
+		return false, fmt.Errorf("serviceAttachment desired target service URL, %q, is malformed: %w", desiredSA.TargetService, err)
 	}
 	if !reflect.DeepEqual(existingFR, desiredFR) {
-		return fmt.Errorf("serviceAttachment target service cannot be updated from %s to %s", existingSA.TargetService, desiredSA.TargetService)
+		return true, fmt.Errorf("serviceAttachment forwarding rule cannot be updated ")
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: good idea. fixed.

File: pkg/psc/controller.go [hunk 13]
```diff
@@ -600,52 +618,67 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
+	// Since forwarding rules and subnets are the same, set them on the desiredCopy to be able to
+	// compare the rest of the fields.
+	desiredCopy := &beta.ServiceAttachment{}
+	*desiredCopy = *desiredSA
+	desiredCopy.TargetService = existingSA.TargetService
+	desiredCopy.NatSubnets = existingSA.NatSubnets
+	// Set region to avoid selflink mismatches
+	desiredCopy.Region = existingSA.Region
+
+	if reflect.DeepEqual(desiredCopy, existingSA) {
```
[the remark this answers is not in the document]
Comment (swetharepakula) [reply pkg/psc/controller.go]: fixed.

**Comment by swetharepakula** [comment]
Newest patch fixes bugs in `needsUpdate` and adds a test case to test that specifically. Some of it overlaps with the `ServiceAttachmentUpdate` test but `needsUpdate` adds additional confidence that we are not making unnecessary GCE API Patch calls.

File: pkg/psc/controller.go [hunk 14]
```diff
@@ -600,52 +618,67 @@ func validateResourceReference(ref v1.TypedLocalObjectReference) error {
...
 
-			for _, subnet := range desiredSA.NatSubnets {
-				desiredSN, err := cloud.ParseResourceURL(subnet)
-				if err != nil {
-					return fmt.Errorf("serviceAttachment desired subnet has malformed URL: %w", err)
-				}
+	for _, desiredSubnet := range desiredSA.NatSubnets {
+		desiredSN, err := cloud.ParseResourceURL(desiredSubnet)
+		if err != nil {
+		}
```
Comment (prameshj) [inline pkg/psc/controller.go:661]: return here

File: pkg/psc/controller_test.go [hunk 15]
```diff
@@ -531,30 +548,136 @@ func TestServiceAttachmentUpdate(t *testing.T) {
...
 			updatedSA, err := getServiceAttachment(controller.cloud, gceSAName)
 			if err != nil {
-				t.Fatalf("Unexpected error when getting updatd GCE ServiceAttachment: %q", err)
+				t.Fatalf("Unexpected error when getting updated GCE ServiceAttachment: %q", err)
 			}
 
-			if !reflect.DeepEqual(createdSA, updatedSA) {
-				t.Errorf("GCE Service Attachment should not be updated. \nOriginal SA:\n %+v, \nUpdated SA:\n %+v", createdSA, updatedSA)
+			if !reflect.DeepEqual(updatedSA, expectedSA) {
+				t.Errorf("Incorrect GCE Service Attachment \nWanted SA:\n %+v, \n got:\n %+v", expectedSA, updatedSA)
```
Comment (prameshj) [inline pkg/psc/controller_test.go:578]: nit: print Got before want. See "Got before Want" in https://github.com/golang/go/issues/34384

File: pkg/psc/controller_test.go [hunk 16]
```diff
@@ -531,30 +548,136 @@ func TestServiceAttachmentUpdate(t *testing.T) {
...
+	saDiffSpec.EnableProxyProtocol = false
+
+	saNoChange := &beta.ServiceAttachment{}
+	*saNoChange = *originalSA
+
+	testcases := []struct {
+		desc           string
+		newSA          *beta.ServiceAttachment
+		expectError    bool
+		expectedResult bool
```
Comment (prameshj) [inline pkg/psc/controller_test.go:636]: nit: maybe call this expectUpdate?

File: pkg/psc/controller.go (hunk 14 above)
Comment (swetharepakula) [reply pkg/psc/controller.go:661]: good catch. The error got deleted by accident. thanks!

File: pkg/psc/controller_test.go (hunk 15 above)
Comment (swetharepakula) [reply pkg/psc/controller_test.go]: fixed

File: pkg/psc/controller.go (hunk 14 above)
Comment (swetharepakula) [reply pkg/psc/controller.go:662]: fixed

File: pkg/psc/controller_test.go (hunk 16 above)
Comment (swetharepakula) [reply pkg/psc/controller_test.go]: fixed

**Pull request merged**